pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+    """
+    A generator that yields a sequence of nodes
+    in an expression tree that belong to a specified set.
+
+    Args:
+        expr: The root node of an expression tree.
+        component_types (set or list): A set of class
+            types that will be matched during the search.
+
+    Yields:
+        Each node that is found.
+    """
+    #
+    # OPTIONS:
+    # component_types - set (or list) if class types to find
+    # in the expression.
+    #
+    visitor = _ComponentVisitor(component_types)
+    for v in visitor.xbfs_yield_leaves(expr):
+        yield v
+
+
+# =====================================================
+#  identify_variables
+# =====================================================
+
+class _VariableVisitor(SimpleExpressionVisitor):
+
+    def __init__(self):
+        self.seen = set()
+
+    def visit(self, node):
+        if node.__class__ in nonpyomo_leaf_types:
+            return
+
+        if node.is_variable_type():
+            if id(node) in self.seen:
+                return
+            self.seen.add(id(node))
+            return node
+
+
+def identify_variables(expr, include_fixed=True):
+    """
+    A generator that yields a sequence of variables
+    in an expression tree.
+
+    Args:
+        expr: The root node of an expression tree.
+        include_fixed (bool): If :const:`True`, then
+            this generator will yield variables whose
+            value is fixed.  Defaults to :const:`True`.
+
+    Yields:
+        Each variable that is found.
+    """
+    visitor = _VariableVisitor()
+    if include_fixed:
+        for v in visitor.xbfs_yield_leaves(expr):
+            yield v
+    else:
+        for v in visitor.xbfs_yield_leaves(expr):
+            if not v.is_fixed():
+                yield v
+
+
+# =====================================================
+#  identify_mutable_parameters
+# =====================================================
+
+class _MutableParamVisitor(SimpleExpressionVisitor):
+
+    def __init__(self):
+        self.seen = set()
+
+    def visit(self, node):
+        if node.__class__ in nonpyomo_leaf_types:
+            return
+
+        # TODO: Confirm that this has the right semantics
+        if not node.is_variable_type() and node.is_fixed():
+            if id(node) in self.seen:
+                return
+            self.seen.add(id(node))
+            return node
+
+
+def identify_mutable_parameters(expr):
+    """
+    A generator that yields a sequence of mutable
+    parameters in an expression tree.
+
+    Args:
+        expr: The root node of an expression tree.
+
+    Yields:
+        Each mutable parameter that is found.
+    """
+    visitor = _MutableParamVisitor()
+    for v in visitor.xbfs_yield_leaves(expr):
+        yield v
+
+
+# =====================================================
+#  _polynomial_degree
+# =====================================================
+
+class _PolyDegreeVisitor(ExpressionValueVisitor):
```

Comment:
For consistency, this should be `_PolynomialDegreeVisitor`

Reply:
Agreed